algebra/src/curves/check_curve_parameters.sage

```diff
@@ -165,29 +167,57 @@ if Fr(COFACTOR) * Fr(COFACTOR_INV) == Fr(SCALAR_FIELD_R):
-## The Halo paper (https://eprint.iacr.org/2019/1021.pdf) claims the injectivity of the endo_mul map.
+####### Checking the consistency of ENDO_COEFF and ENDO_SCALAR #############
+if endo_mul_is_used:
+    Q = int(zeta_r) * P
+    if Q == E([zeta_q * X, Y]):
+        print("Correct. ENDO_COEFF and ENDO_SCALAR are consistent.")
+    else:
+        print("WARNING! ENDO_COEFF AND ENDO_SCALAR ARE NOT CONSISTENT!")
+
+
+########## Checking that shortest vector in the lattice ([1,zeta_r),[0,r]) is long enough #########
+## The Halo paper (https://eprint.iacr.org/2019/1021.pdf) proves the injectivity of the endo_mul map.
 ## The injectivity of the map (a,b) |-> a\zeta + b for a,b in [0,A] (essential for using add_unsafe)
 ## is equivalent the lattice condition below.
-## TODO: Write more in detail the motivation of this equivalence (not done in this way in the paper).
-#### Bound for a,b in endo_mul, see the Halo paper:
-## In general it is 2**(l/2 + 1) + 2**(l/2), where l is the length of the input string (in endo_mul we use l = 128)
-A = 2**65 + 2**64 
-from sage.modules.free_module_integer import IntegerLattice
-L_q = IntegerLattice([[1,int(zeta_q)],[0,BASE_FIELD_MODULUS]])
-if L_q.shortest_vector().norm().n() <= (sqrt(2)*A).n():
-    print("WARNING! WE CAN'T USE add_unsafe FOR endo_mul")
-L_r = IntegerLattice([[1,int(zeta_r)],[0,SCALAR_FIELD_MODULUS]])
-if L_r.shortest_vector().norm().n() <= (sqrt(2)*A).n():
-    print("WARNING! WE CAN'T USE add_unsafe FOR endo_mul")
\ No newline at end of file
+## In the Halo paper is explained that the condition that a*zeta + b != a'*zeta + b' for a,a',b,b' in [0,A] is equivalent
+## to the condition that min({distance(zeta*a,zeta*a'): a,a' in [0,A]}) > A. This can be plainly restated as
+## a*zeta notin [-A,A] mod r if a in [-A,A] mod r.
+## This can be restated as a lattice condition: suppose in fact that
+##      a * zeta = b mod r  for a,b in [-A,A].
+## Then it would exists c such that
+##      b = a * zeta + c * r.
+## Then
+##      (a, b) = (a, c) * (1  zeta)
+##                        (0    r )
+## would be vector of length <= sqrt(2)*A in a lattice of  determinant r (Gaussian expectation for the shortest vector ~sqrt(r)).
+## If we show that the shortest vector in the lattice spanned by (1, zeta) and (0, r) is longer than sqrt(2)*A, then we
+## have a sufficient (but not necessary) condition.
+## We could have indeed also a necessary condition: if there were a Sage functions listing all the vector shorter than sqrt(2)*A
+## one can look at them: if none of them has both coordinates in the range [-A,A] then the parameters can be accepted
+## even if the shortest vector has length in the range (A, sqrt(2)A].
+## Anyway, due to our choice of parameters, A ~ 2^65, while sqrt(r) ~ 2^128.
+## In general it is A = 2**(l/2 + 1) + 2**(l/2), where l is the length of the input string (in endo_mul we use l = 128)
```

Comment:
As discussed offline, unless it is easy to modify the lattice based argument: let us implement the Euclidian algorithm test, which serves a sufficient AND necessary condition for the injectivity of the endo_rep to scalar map.